griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import annotations
+
+import json
+import os
+from typing import Optional
+
+from attrs import Factory, define, field
+
+from griptape.artifacts import BaseArtifact
+from griptape.clients import GriptapeCloudApiClient
+from griptape.drivers import BaseConversationMemoryDriver
+from griptape.memory.structure import BaseConversationMemory, ConversationMemory, Run
+from griptape.mixins import ExponentialBackoffMixin
+
+
+@define(kw_only=True)
+class GriptapeCloudConversationMemoryDriver(BaseConversationMemoryDriver, ExponentialBackoffMixin):
+    id: str = field(
+        default=Factory(lambda self: os.getenv("GT_CLOUD_THREAD_ID", self._get_id()), takes_self=True),
+        metadata={"serializable": True},
+    )
+    api_key: str = field(default=None, metadata={"serializable": False})
+    base_url: str = field(default=None, metadata={"serializable": True})
+    structure_run_id: str = field(default=None, metadata={"serializable": True})
+    client: GriptapeCloudApiClient = field(
+        default=Factory(
+            lambda self: GriptapeCloudApiClient(
+                api_key=self.api_key, base_url=self.base_url, structure_run_id=self.structure_run_id
+            ),
+            takes_self=True,
+        ),
+        metadata={"serializable": False},
+    )
+
+    def _get_id(self) -> Optional[str]:
+        thread_id = None
+        for attempt in self.retrying():
+            with attempt:
+                thread_id = self.client.post("/threads", {}).get("thread_id")
+
+        return thread_id
+
+    def store(self, memory: BaseConversationMemory) -> None:
+        messages = [{"input": json.dumps(run.input), "output": json.dumps(run.output)} for run in memory.runs]
+        for attempt in self.retrying():
+            with attempt:
+                self.client.patch(f"/threads/{self.id}", {"messages": messages})
+
+    def load(self) -> Optional[BaseConversationMemory]:
+        response = None
+        for attempt in self.retrying():
+            with attempt:
+                response = self.client.get(f"/threads/{self.id}/messages")
+        if response is None:
+            return None
+        messages: list[dict] = response.get("messages", [])
+        messages = sorted(messages, key=lambda m: m["index"])
+        runs = [
+            Run(
+                id=m["message_id"], input=BaseArtifact.from_dict(m["input"]), output=BaseArtifact.from_dict(m["output"])
+            )
+            for m in messages
+        ]
+        return ConversationMemory(runs=runs, driver=self)
```

Comment:
Needs to support other Conversation Memory types like `SummaryConversationMemory`. Fields like `SummaryConversationMemory.summary` should be stored in thread metadata.